pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/mutablefst/MutableFST.java

```diff
@@ -0,0 +1,66 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.local.utils.nativefst.mutablefst;
+
+
+/**
+ * A mutable FST represents a FST which can have arbitrary inputs added to it at
+ * any given point of time. Unlike a normal FST which is build once read many, mutable
+ * FST can be concurrently written to and read from. Mutable FST provides real time search
+ * i.e. search will see words as they are added without needing a flush.
+ *
+ * Unlike a normal FST, mutable FST does not require the entire input beforehand nor
+ * does it require the input to be sorted. Single word additions work well with
+ * mutable FST.
+ *
+ * The reason as to why normal FST and mutable FST have different interfaces is because
```

Comment:
Does this mean that there will have to be separate code to use mutable vs immutable FST? An abstraction that avoids that would be great.

Reply:
> Thanks a lot for adding this powerful feature. It would be great to see a proposal on high this ties in end-to-end (as in would the interfaces still hold).

Added in the documentation

Reply:
> Does this mean that there will have to be separate code to use mutable vs immutable FST? An abstraction that avoids that would be great.

That's a great point. However, here is the catch. Immutable and Mutable FSTs, while providing the same functionality, follow different data structures. It is pretty hard to get them to agree on one interface without significantly compromising ImmutableFST's serialised state optimisations or Mutable FST's speed of addition of new paths.

However, any layer above pinot-segment will not be aware of this difference. Referring to the proposal document above, the integration path for mutable FST will include a native FST reader which will propagate a query to both mutable and immutable FSTs(within a segment) at the same time.